### ISSUE TITLE
[YOLO26-cls ONNX Runtime inference outputs inconsistent classification results compared with official PyTorch predict()

### ISSUE THREAD

**Issue by zsffuture** [ultralytics/ultralytics#25794]
### Search before asking

- [x] I have searched the Ultralytics YOLO [issues](https://github.com/ultralytics/ultralytics/issues) and [discussions](https://github.com/orgs/ultralytics/discussions) and found no similar questions.

### Question

Title
[YOLO26-cls ONNX Runtime inference outputs inconsistent classification results compared with official PyTorch predict()
Body
Problem description
I exported YOLO26 classification ONNX model (weights FP16, graph contains Softmax inside).
When deploying via ONNX Runtime with Python, the predicted class results are inconsistent with the official model.predict() from ultralytics.
Preprocessing pipeline (strictly following official training transform):
Resize: equal scale resize, short side to 640, bilinear interpolation, antialias=True
CenterCrop to [640, 640]
ToTensor()
Normalize(mean=[0.,0.,0.], std=[1.,1.,1.])
Key details:
Input tensor feed to ONNX is float32 (graph input defined as tensor(float), model internal weights FP16)
No manual softmax in code, ONNX graph already includes Softmax layer
Preprocessing implemented with pure numpy/PIL, replicated the exact logic of torchvision center_crop source code
Channel order: RGB, image range [0, 1] after ToTensor
Environment
ultralytics: latest main branch
onnxruntime-gpu
PIL for image processing
Question
What possible factors cause the classification mismatch between ONNX runtime deployment and native PyTorch inference?
Is there any hidden difference in resize rounding rule, interpolation boundary handling or datatype precision between PyTorch and ONNX exported model?

**Comment by UltralyticsAssistant** [comment]
👋 Hello @zsffuture, thank you for reporting the inconsistent YOLO26 classification results between ONNX Runtime and native PyTorch inference. The preprocessing details are helpful, but a comparison script will be needed to determine whether the difference comes from preprocessing, export, or runtime behavior.

Please provide a [minimum reproducible example](https://docs.ultralytics.com/help/minimum-reproducible-example) including:
- The exact ONNX export command and model checkpoint used
- Exact `ultralytics`, PyTorch, ONNX Runtime, Python, and hardware/CUDA versions
- The PyTorch `model.predict()` call and ONNX Runtime inference code
- One input image, or a reproducible way to obtain it
- The raw ONNX output and PyTorch class probabilities for that same image, including output shapes
- The `imgsz`, device, and confidence-related settings used for the PyTorch prediction

Please also upgrade to the latest `ultralytics` package and verify whether the mismatch persists. This is an automated response; an Ultralytics engineer will assist soon.

**Comment by zsffuture** [comment]
导出代码：
model = YOLO("runs\\classify\\train\\weights\\last.pt")
    model.export(
        format="onnx",
        dynamic=False,  # (1)! 

        imgsz=(640, 640),
        simplify=True,
        quantize=16,
       
    )

**Comment by zsffuture** [comment]
ultralytics  infer  results:
Top-1 class: 哑光灰_UK, Confidence: 0.9849
Top-5 class: 哑光灰_UK, Confidence: 0.9849
Top-5 class: 航空银_KX, Confidence: 0.0150
Top-5 class: 月光银_KU, Confidence: 0.0000
Top-5 class: 薄雾绿_SY, Confidence: 0.0000
Top-5 class: 碳晶黑_CL, Confidence: 0.0000
my code infer:
===== 分类结果 =====
cls_航空银_KX : 0.9850
cls_哑光灰_UK : 0.0150
cls_卡其白_BW : 0.0000
cls_月光银_KU : 0.0000
cls_碳晶黑_CL : 0.0000

my code :
import cv2
import numpy as np
import onnxruntime as ort
from PIL import Image


def get_dimensions_np(img: np.ndarray):
    """模拟torchvision get_dimensions, CHW格式 [C,H,W]"""
    c, h, w = img.shape
    return c, h, w


def center_crop_np(img: np.ndarray, output_size):
    """
    完全复刻官方torchvision center_crop逻辑
    :param img: np.ndarray CHW RGB float32
    :param output_size: [H, W]
    :return: cropped CHW array
    """
    import numbers
    if isinstance(output_size, numbers.Number):
        output_size = (int(output_size), int(output_size))
    elif isinstance(output_size, (tuple, list)) and len(output_size) == 1:
        output_size = (output_size[0], output_size[0])
    crop_h, crop_w = output_size
    _, img_h, img_w = get_dimensions_np(img)

    # 需要padding场景
    pad_ltrb = [0, 0, 0, 0]  # left, top, right, bottom
    need_pad = False
    if crop_w > img_w:
        dl = (crop_w - img_w) // 2
        dr = (crop_w - img_w + 1) // 2
        pad_ltrb[0] = dl
        pad_ltrb[2] = dr
        need_pad = True
    if crop_h > img_h:
        dt = (crop_h - img_h) // 2
        db = (crop_h - img_h + 1) // 2
        pad_ltrb[1] = dt
        pad_ltrb[3] = db
        need_pad = True

    if need_pad:
        # np.pad 格式 (C维不pad, H上下pad, W左右pad)
        pad_width = ((0, 0), (pad_ltrb[1], pad_ltrb[3]), (pad_ltrb[0], pad_ltrb[2]))
        img = np.pad(img, pad_width, mode="constant", constant_values=0)
        _, img_h, img_w = get_dimensions_np(img)
        if crop_w == img_w and crop_h == img_h:
            return img

    # 中心裁剪坐标
    crop_top = int(round((img_h - crop_h) / 2.0))
    crop_left = int(round((img_w - crop_w) / 2.0))
    return img[:, crop_top:crop_top + crop_h, crop_left:crop_left + crop_w]


def preprocess_image(image, target_side=640, crop_size=(640, 640), use_short_side=True):
    """
    和训练预处理完全对齐
    :param image: 图片路径 / cv2 BGR数组
    :param target_side: resize对齐边长
    :param crop_size: (H,W)
    :param use_short_side: True=短边缩放到target_side(匹配你示例2808*2196→844*640)
                           False=长边缩放到target_side
    :return: input tensor shape [1,3,640,640] np.float16 适配FP16 ONNX
    """
    if isinstance(image, str):
        pil_img = Image.open(image).convert("RGB")
    else:
        # cv2 BGR -> RGB
        rgb = cv2.cvtColor(image, cv2.COLOR_BGR2RGB)
        pil_img = Image.fromarray(rgb)

    w, h = pil_img.size

    # 等比例缩放
    if use_short_side:
        scale = target_side / min(h, w)
    else:
        scale = target_side / max(h, w)
    new_w = int(round(w * scale))
    new_h = int(round(h * scale))

    # bilinear + antialias=True 对应PIL Image.Resampling.BILINEAR
    pil_img = pil_img.resize((new_w, new_h), Image.Resampling.BILINEAR)

    # HWC uint8 -> CHW float32 [0,1] (ToTensor)
    arr = np.array(pil_img, dtype=np.float32) / 255.0
    arr = np.transpose(arr, (2, 0, 1))  # HWC -> CHW

    # center crop
    arr = center_crop_np(arr, crop_size)

    # normalize mean [0,0,0], std[1,1,1] 无需运算
    # 增加batch维度 + 转FP16适配onnx fp16输入
    input_tensor = np.expand_dims(arr, axis=0).astype(np.float32)
    return input_tensor


class YOLO26ClsONNX:
    def __init__(self, onnx_path: str, class_names: list, use_gpu=True):
        self.class_names = class_names
        providers = []
        if use_gpu:
            providers.append("CUDAExecutionProvider")
        providers.append("CPUExecutionProvider")
        self.session = ort.InferenceSession(onnx_path, providers=providers)
        self.input_name = self.session.get_inputs()[0].name
        self.output_name = self.session.get_outputs()[0].name
        print("ONNX模型加载完成")

    def predict(self, img, topk=5):
        tensor = preprocess_image(img, target_side=640, crop_size=(640, 640), use_short_side=True)
        outputs = self.session.run([self.output_name], {self.input_name: tensor})
        # ONNX内部已包含Softmax，output直接是概率
        probs = outputs[0][0].astype(np.float32)

        top_idx = np.argsort(probs)[::-1][:topk]
        res = [(self.class_names[i], float(probs[i])) for i in top_idx]
        return res


if __name__ == "__main__":
    # ===================== 配置区 =====================
    ONNX_MODEL = "runs\\classify\\train\\weights\\yolo26l-cls.onnx" 
    # ImageNet1000类别，如果你是自定义数据集替换此处列表
    CLASS_NAMES = [f"cls_{i}" for i in [  "薄雾绿_SY","航空银_KX","卡其白_BW","碳晶黑_CL","哑光灰_UK","月光银_KU"]]
    IMG_PATH = "runs\\classify\\1.jpg"
    # ==================================================
    classifier = YOLO26ClsONNX(
        onnx_path=ONNX_MODEL,
        class_names=CLASS_NAMES,
        use_gpu=False  # 如果有CUDA环境可以改为True
    )
    img = cv2.imread(IMG_PATH)
    res = classifier.predict(img, topk=5)
    print("===== 分类结果 =====")
    for name, conf in res:
        print(f"{name} : {conf:.4f}")

**Comment by zsffuture** [comment]
ok ,i know 
I have identified the cause. Classification tasks do not rely on a YAML configuration file — class names are fetched directly from the dataset list. I was instead using my self-defined class name list, and this discrepancy caused the inconsistent prediction results. id is same

image 1/1 E:\vscode_prj\train_prj\jiaPian\runs\classify\1.jpg: 640x640 _UK 0.98, _KX 0.01, _KU 0.00, _SY 0.00, _CL 0.00, 16.2ms
Speed: 34.8ms preprocess, 16.2ms inference, 0.2ms postprocess per image at shape (1, 3, 640, 640)
Top-1 class: 哑光灰_UK, Confidence: 0.9849
Top-5 class: 哑光灰_UK, Confidence: 0.9849
Top-5 class: 航空银_KX, Confidence: 0.0150
Top-5 class: 月光银_KU, Confidence: 0.0000
Top-5 class: 薄雾绿_SY, Confidence: 0.0000
Top-5 class: 碳晶黑_CL, Confidence: 0.0000
id: [1, 4, 2, 5, 3]
(base) PS E:\vscode_prj\train_prj\jiaPian>  e:; cd 'e:\vscode_prj\train_prj\jiaPian'; & 'D:\install\anaconda3\python.exe' 'c:\Users\zsffuture\.vscode\extensions\ms-python.debugpy-2026.6.0-win32-x64\bundled\libs\debugpy\launcher' '53434' '--' 'E:\vscode_prj\train_prj\jiaPian\infer_classify.py' 
top_idx: [1 4 2 5 3]
===== 分类结果 =====
cls_航空银_KX : 0.9850
cls_哑光灰_UK : 0.0150
cls_卡其白_BW : 0.0000
cls_月光银_KU : 0.0000
cls_碳晶黑_CL : 0.0000
